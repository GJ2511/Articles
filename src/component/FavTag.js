import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

function FavTag({ handleTagClick, selectedTag, tags, setSeletedTag }) {
    const handleClick = (selected, tag) => {
        setSeletedTag(selected, tag);
        handleTagClick();
    };

    return tags.map((tag, index) => {
        const selected = tag === selectedTag ? true : false;
        const classes = cx('badge ml-1', {
            'badge-secondary': !selected,
            'badge-primary': selected,
        });

        return (
            <span className={classes} key={index} onClick={() => handleClick(selected, tag)}>
                {tag}
            </span>
        );
    });
}

FavTag.propTypes = {
    handleTagClick: PropTypes.func.isRequired,
    selectedTag: PropTypes.string,
    tags: PropTypes.array.isRequired,
};

export default FavTag;
