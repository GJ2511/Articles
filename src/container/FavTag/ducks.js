import { takeLatest, call, put } from 'redux-saga/effects';

import TagService from '../../services/tagService';

const initialState = {
    loading: true,
    tags: [],
    selectedTag: '',
};

const PREFIX = 'FAV_TAG';
const FAV_TAG_REQUESTED = `${PREFIX}//FAV_TAG_REQUESTED`;
const FAV_TAG_SUCCESS = `${PREFIX}//FAV_TAG_SUCCESS`;
const FAV_TAG_ERROR = `${PREFIX}//FAV_TAG_ERROR`;
const SET_SELECTED_TAG = `${PREFIX}//SET_SELECTED_TAG`;

const favTagReducer = (state = initialState, action = {}) => {
    const { type, payload } = action;

    switch (type) {
        case FAV_TAG_SUCCESS:
            return {
                ...state,
                loading: false,
                tags: payload,
            };
        case FAV_TAG_ERROR:
            return {
                ...state,
                loading: false,
                tags: [],
            };
        case SET_SELECTED_TAG:
            return {
                ...state,
                selectedTag: payload,
            };
        default:
            return state;
    }
};

export const favTagRequested = () => ({ type: FAV_TAG_REQUESTED });
export const setSeletedTag = (isSelected, val) => ({ type: SET_SELECTED_TAG, payload: !isSelected ? val : '' });
export const getSeletedTag = (state) => state.favTagReducer.selectedTag;

function* favTag() {
    try {
        const response = yield call([TagService, 'getTags']);

        if (response.errors) {
            yield put({ type: FAV_TAG_ERROR });
        } else {
            yield put({ type: FAV_TAG_SUCCESS, payload: response.tags });
        }
    } catch (error) {
        yield put({ type: FAV_TAG_ERROR });
    }
}

export function* favTagSaga() {
    yield takeLatest(FAV_TAG_REQUESTED, favTag);
}

export const actions = {
    favTagRequested,
    setSeletedTag,
};

export default favTagReducer;
