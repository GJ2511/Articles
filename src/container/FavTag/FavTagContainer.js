import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Loader from '../../component/Loader';
import FavTag from '../../component/FavTag';
import { favTagRequested, setSeletedTag } from './ducks';

class FavTagContainer extends Component {
    componentDidMount() {
        this.props.favTagRequested();
    }

    render() {
        const { loading, ...rest } = this.props;

        return (
            <div className="card card-signin bg-light">
                <div className="card-header">
                    <h5 className="card-title text-center">Popular Tags</h5>
                </div>
                <div className="card-body bg-white">{loading ? <Loader /> : <FavTag {...rest} />}</div>
            </div>
        );
    }
}

FavTagContainer.propTypes = {
    favTagRequested: PropTypes.func.isRequired,
    handleTagClick: PropTypes.func.isRequired,
    loading: PropTypes.bool.isRequired,
    selectedTag: PropTypes.string,
    setSeletedTag: PropTypes.func.isRequired,
    tags: PropTypes.array.isRequired,
};

const mapStateToProps = ({ favTagReducer }) => ({
    tags: favTagReducer.tags,
    loading: favTagReducer.loading,
    selectedTag: favTagReducer.selectedTag,
});

export default connect(mapStateToProps, { favTagRequested, setSeletedTag })(FavTagContainer);
