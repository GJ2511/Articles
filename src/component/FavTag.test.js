import React from 'react';

import FavTag from './FavTag';

const props = {
    handleTagClick: jest.fn(),
    selectedTag: 'reactjs',
    tags: ['reactjs', 'angularjs'],
};

describe('<FavTag tests />', () => {
    test('match snapshot', () => {
        expect(shallow(<FavTag {...props} />)).toMatchSnapshot();
    });
});
