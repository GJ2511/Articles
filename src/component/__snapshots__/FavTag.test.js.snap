// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FavTag tests /> match snapshot 1`] = `
Array [
  <span
    className="badge ml-1 hover-hand badge-primary"
    key="0"
    onClick={[Function]}
  >
    reactjs
  </span>,
  <span
    className="badge ml-1 hover-hand badge-secondary"
    key="1"
    onClick={[Function]}
  >
    angularjs
  </span>,
]
`;
